src/java.base/share/classes/java/util/ServiceLoader.java

```diff
@@ -884,6 +893,16 @@ private Provider<S> loadProvider(ServiceProvider provider) {
 
                 @SuppressWarnings("unchecked")
                 Class<? extends S> type = (Class<? extends S>) returnType;
+                // If the restricted security mode is enabled
+                // && the specified class extends java.security.Provider
+                // && the provider is NOT allowed in restricted security mode.
+                if (RestrictedSecurityConfigurator.isEnabled()
+                        && java.security.Provider.class.isAssignableFrom(clazz)
+                        && !RestrictedSecurityProperties.getInstance()
+                                .isProviderAllowed(clazz.getName())) {
```

Comment:
@keithc-ca 

In the ProviderConfig.java, it will also call this isProviderAllowed() method, to check if the java.base provider is allowed or not. Because the java.base provider will be initialized directly, not by the service loader. And at that place, it use the provider name, not the class.

So, I added an overload method with Class object as the parameter in the RestrictedSecurityProperties.java, to first check if the class extends java.security.Provider or not. If not, return false. If yes, call the isProviderAllowed() method with String as parameter to continue check if the provider is allowed or not.

Reply:
@keithc-ca 
Grinder openjdk tests failed after moving the isAssignable() into the isProviderAllowed() method. Working on it and will update shortly, mark this unresolved again.